### DISCUSSION TITLE
How to pass conversation history and implement chat-like experience?

### DISCUSSION THREAD

**Discussion by remixtedi** [KeyserDSoze/Rystem.OpenAi#46] [Q&A]
I couldn't find how to save and pass the conversation history between the user and the assistant to create a chat-like experience.
Is it possible with this library? can you provide an example if it exists?

**Comment by KeyserDSoze** [comment]
You have to do it by yourself. You need to recreate everytime all the chat history. Here an example

```
var chatHistory = openAiApi.Chat.RequestWithSystemMessage("for example something to pass to the system.")
                .AddUserMessage("first message from the user");
            var message = await chatHistory
                .ExecuteAsync();
            //here the first output for the user from openai
            var contentFromAssistant = message.Choices[0].Message.Content;
            //response from openai added to the chat builder
            chatHistory.AddAssistantMessage(contentFromAssistant);
            //message from the user
            chatHistory.AddUserMessage("second message from the user");
            var message2 = await chatHistory
               .ExecuteAsync();
            //here the second output for the user from openai
            var secondContentFromAssistant = message2.Choices[0].Message.Content;
            //response from openai added to the chat builder
            chatHistory.AddAssistantMessage(contentFromAssistant);
            //message from the user
            chatHistory.AddUserMessage("third message from the user");
            var message3 = await chatHistory
               .ExecuteAsync();
            //and so on......
```

**Reply by remixtedi** [reply]
I created a foreach loop and it worked but it would be much simple and clear to have a function that can receive a collection of ChatMessage objects like openAIApi.Chat.AddMessages(collectionOfChatMessages);

**Reply by KeyserDSoze** [reply]
ok it sounds good, i'm going to add it in next version (probably today night)

**Reply by KeyserDSoze** [reply]
Something like this?
![image](https://github.com/KeyserDSoze/Rystem.OpenAi/assets/40713438/b3482525-1c7d-472d-b7e8-adddae746f72)

**Reply by KeyserDSoze** [reply]
Furthermore I would add the chance to retrieve all messages inserted 
![image](https://github.com/KeyserDSoze/Rystem.OpenAi/assets/40713438/4b12fb01-5046-4482-8bca-978a9d469c52)

**Comment by remixtedi** [comment]
@KeyserDSoze Looks good <3

**Comment by nluo1201** [comment]
I tried to solve this issue. The current setup does not allow a quick and easy way to control over the reconstruction of message history that has different ordering of message roles. For example, sometime you want to start with no system message, a user role message, or the assistant message itself.
![image](https://github.com/KeyserDSoze/Rystem.OpenAi/assets/10542843/af4917fe-0c1f-450d-a1fc-2cd81588a7b0)

**Reply by nluo1201** [reply]
@KeyserDSoze Any issues you can think of with this additions?

**Reply by KeyserDSoze** [reply]
Ok I think tomorrow I can implement this issue.

**Reply by KeyserDSoze** [reply]
Now you can start without a message
![image](https://github.com/KeyserDSoze/Rystem.OpenAi/assets/40713438/1a759269-f4d9-46ab-a3df-4fb3e7d34591)
